### PR TITLE
Code modernising

### PR DESCRIPTION
**Pull request by Lakritzator**
These changes should improve the readability and stability of the code base, also when moving / modifying. Especially not using hard-coded strings.

As the project was already at C# 7.3, I modified some of the code to use more of the features, but only if these improve readability. I could have been a lot more aggressive, but assume that would be to much.

Removed the "this." references where possible, as it's unneeded from my standpoint.
Moved 2 classes in their own files, it's a common convention not to have multiple classes in one file.

In TaskbarIcon there was a lot of lock(this), which is bad practice (but most likely not an issue here), I replaced this with a lock on a separate object. This might increase the memory usage with a few bytes, but I would advice to do it anyway.

Also fixed some typos in documentation.

What is missing, are some code / naming conventions (like are private fields prefixed with an _ or not etc)

Just look if you like these changes

### CODE REVIEW

**Comment by punker76** [comment]
@hardcodet LGTM 👍 @Lakritzator

File: Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/Interop/BalloonFlags.cs [hunk 1]
```diff
@@ -1,9 +1,12 @@
+using System;
+
 namespace Hardcodet.Wpf.TaskbarNotification.Interop
 {
     /// <summary>
     /// Flags that define the icon that is shown on a balloon
     /// tooltip.
     /// </summary>
+    [Flags]
```
Comment (hardcodet) [inline Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/Interop/BalloonFlags.cs]: What is your reason to add the `Flags` attribute? You cannot combine them, after all :)

File: Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/Interop/SystemInfo.cs [hunk 2]
```diff
@@ -2,37 +2,41 @@
...
-      }
-    }         
-  }
-}
+        public static double DpiXFactor
+        {
+            get
+            {
+                var factors = DpiFactors;
+                return factors?.X ?? 1;
```
Comment (hardcodet) [inline Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/Interop/SystemInfo.cs]: Wouldn't sign off on this being more readable than the old-school evaluation, but fine by me 👍

File: Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/Interop/BalloonFlags.cs (hunk 1 above)
Comment (Lakritzator) [reply Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/Interop/BalloonFlags.cs]: But you are: https://github.com/hardcodet/wpf-notifyicon/blob/master/Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/TaskbarIcon.cs#L814

File: Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/Interop/SystemInfo.cs (hunk 2 above)
Comment (Lakritzator) [reply Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/Interop/SystemInfo.cs]: I actually considered changing more here, as the name of the class somehow doesn't represent the functionality it provides, and the code can be simplified.

What about this:

```
    /// <summary>
    /// This class is a helper to get the DPI factors
    /// </summary>
    public static class DpiFactor
    {
        private static readonly System.Windows.Point DpiFactors;

        static DpiFactor()
        {
            using (var source = new HwndSource(new HwndSourceParameters()))
            {
                if (source.CompositionTarget?.TransformToDevice != null)
                {
                    DpiFactors = new System.Windows.Point(source.CompositionTarget.TransformToDevice.M11, source.CompositionTarget.TransformToDevice.M22);
                    return;
                }
                DpiFactors = new System.Windows.Point(1, 1);
            }
        }

        /// <summary>
        /// Returns the DPI X Factor
        /// </summary>
        public static double FactorX => DpiFactors.X;

        /// <summary>
        /// Returns the DPI Y Factor
        /// </summary>
        public static double FactorY => DpiFactors.Y;
    }

```

Comment (Lakritzator) [reply Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/Interop/SystemInfo.cs]: The calling code would be:

`return new Point { X = (int)(point.X / DpiFactor.FactorX), Y = (int)(point.Y / DpiFactor.FactorY) };`

Instead of:
`return new Point() { X = (int)(point.X / SystemInfo.DpiXFactor), Y = (int)(point.Y / SystemInfo.DpiYFactor) };`

Comment (Lakritzator) [reply Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/Interop/SystemInfo.cs]: Let me know if you like that more, I will push another commit.

**Comment by Lakritzator** [comment]
BTW: I was "struggling" with applying "var", and didn't, I'm not a fan of var everywhere but it has it usages... And it's already used in some places...

File: Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/Interop/BalloonFlags.cs (hunk 1 above)
Comment (hardcodet) [reply Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/Interop/BalloonFlags.cs]: Damn, I am! Still, I'm not sure this merits the flag - it's not your typical bit field... Thoughts?

File: Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/Interop/SystemInfo.cs (hunk 2 above)
Comment (hardcodet) [reply Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/Interop/SystemInfo.cs]: The syntax is much cleaner. As far as the class name goes, it's debatable. From a purist view, your variant has the higher semantic cohesion.

OTOH, I would argue that the `SystemInfo` class name sort of documents that this is an attribute of the system, while the class in itself doesn't provide that context (it's the DPI factor of what?). Looking at this code the first time in years, this actually helped myself getting my head around it.

Bottom line: +1 for the cleaner properties, but I'd keep the class name. If you disagree, convince me :)

File: Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/Interop/BalloonFlags.cs (hunk 1 above)
Comment (punker76) [reply Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/Interop/BalloonFlags.cs]: The Flags attribute is not necessary if you have correct values... https://stackoverflow.com/a/8480/920384

**Comment by hardcodet** [comment]
> 
> 
> BTW: I was "struggling" with applying "var", and didn't, I'm not a fan of var everywhere but it has it usages... And it's already used in some places...

@Lakritzator  ACK, I use var if I can infer the type from looking at the code, otherwise I don't:

```
var user = new User("john");
User user = await GetUser("john");
```

File: Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/Interop/BalloonFlags.cs (hunk 1 above)
Comment (Lakritzator) [reply Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/Interop/BalloonFlags.cs]: I understand your struggle, the first 4 bits are more or less the type and combining them doesn't make sense. But the later values (0x10, 0x20, 0x80) are flags...

It would almost merit a separate struct or class.. :disappointed:

Comment (hardcodet) [reply Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/Interop/BalloonFlags.cs]: I don't think users of the API will ever combine this - plus the attribute would imply that you can arbitrarily combine them (as with your typical `1,2,4,8,...` flags, which isn't the case here. So I'd skip it. OTOH, no point in spending too much time on this, we're probably neck deep in semantics now 😄

File: Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/Interop/SystemInfo.cs (hunk 2 above)
Comment (Lakritzator) [reply Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/Interop/SystemInfo.cs]: I'm not here to convince you :grinning:, the name is okay.

Pushing a commit with keeping the `SystemInfo` name, which forces the property name to contain DPI, but changing the code.

Comment (hardcodet) [reply Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/Interop/SystemInfo.cs]: Perfect! Just drop me a note when you think you're done and I'll merge. Thanks for all the work you put in!

File: Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/Interop/BalloonFlags.cs (hunk 1 above)
Comment (Lakritzator) [reply Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/Interop/BalloonFlags.cs]: Ok so we agree to remove the [Flags]?
I will add a comment for ReSharper to ignore the warning.

Comment (hardcodet) [reply Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/Interop/BalloonFlags.cs]: @Lakritzator  cross-post... yeah, this is a weird one. otoh, it's the windows api, so let's keep together what doesn't really fit together ;)

Comment (hardcodet) [reply Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/Interop/BalloonFlags.cs]: > 
> 
> Ok so we agree to remove the [Flags]?
> I will add a comment for ReSharper to ignore the warning.

I'd either remove it, or keep it with a comment that those can't be arbitrarily combined (may provide more context than just an R# ignore). I'll let you be the judge of it 👍

Comment (Lakritzator) [reply Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/Interop/BalloonFlags.cs]: It's not the weirdest Windows API case :laughing: 

Removed the flag, also found some more comments with typos or missing information.

File: Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/TaskbarIcon.cs [hunk 3]
```diff
@@ -171,21 +175,21 @@ public Point GetPopupTrayPosition()
...
         /// <param name="timeout">The time after which the popup is being closed.
-        /// Submit null in order to keep the balloon open inde
+        /// Submit null in order to keep the balloon open indefinitely
         /// </param>
         /// <exception cref="ArgumentNullException">If <paramref name="balloon"/>
         /// is a null reference.</exception>
         public void ShowCustomBalloon(UIElement balloon, PopupAnimation animation, int? timeout)
         {
-            Dispatcher dispatcher = this.GetDispatcher();
+            var dispatcher = this.GetDispatcher();
```
Comment (hardcodet) [inline Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/TaskbarIcon.cs]: Minor: That for example would be a case where I'd keep it (I don't know what type `GetDispatcher` returns. For all I know, it could return a Task...

Comment (hardcodet) [reply Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/TaskbarIcon.cs]: (Please don't feel you have to commit a correction for this, this is just fine!)

File: Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/TaskbarIcon.cs [hunk 4]
```diff
@@ -156,10 +150,20 @@ public TaskbarIcon()
...
+        /// <summary>
+        /// A delegate to handle customer popup positions.
+        /// </summary>
         public delegate Point GetCustomPopupPosition();
 
-        public GetCustomPopupPosition CustomPopupPosition;
+        /// <summary>
+        /// Specify a custom popup position
+        /// </summary>
+        public GetCustomPopupPosition CustomPopupPosition { get; set; }
```
Comment (hardcodet) [inline Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/TaskbarIcon.cs]: @punker76 @Lakritzator  I assume I had a reason to not declare that as a property, but I haven't done Windows UI for quite a while. Are there any implications to this with regards to data binding?

File: Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/TaskbarIcon.cs (hunk 3 above)
Comment (Lakritzator) [reply Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/TaskbarIcon.cs]: I mildly disagree, a method called `GetDispatcher` should return a Dispatcher and not something which is a completely different thing. Naming is important!

But I have no issue rolling it back, it's your party :1st_place_medal:

File: Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/TaskbarIcon.cs (hunk 4 above)
Comment (Lakritzator) [reply Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/TaskbarIcon.cs]: Do you mean, as long as it's a field it will not be available for Binding? Because currently nothing would be binding to it, which makes sense..

That should be something @punker76 know better.

Actually that would be the only reason I could come up with against changing it, in general it's bad habit to make fields publicly available. (I actually didn't really look for them)

**Comment by Lakritzator** [comment]
Some other things I noticed:
Currently the solution is targeting many framework versions, are these needed? I guess you can reduce it on the essence, like 4.5 & 4.7.1 or so...

I found some sub-optimal interop calls, from a performance & memory perspective, but I think this doesn't need to be changed as it's not a critical path. I will check this another time.

**Comment by Lakritzator** [comment]
As @punker76 is involved, I would expect him to work on Azure Pipeline support too, for making automatic builds & publishing to nuget (if approved), if not I can help.

AppVeyor would be an alternative, but due to the gigantic forward momentum which Microsoft has, my advice would be to use Azure Pipelines.

Your organisation will need an "account" though, but it's free for open source and solves many issues with building locally.

**Comment by hardcodet** [comment]
Yes, this code has been unchanged for years now, it can most certainly use some brushing up with regards to targets. Where possible (and easy!), I'd like to support older framework versions, but given it's been mostly stable, I'm not against cutting off some old version at all.

File: Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/TaskbarIcon.cs (hunk 4 above)
Comment (hardcodet) [reply Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/TaskbarIcon.cs]: I'm not sure, I just have a (six year old ;) deja-vu. Plus I really dislike public fields myself, so there's a bit of a red flag for me. Let's see what Jan thinks about it before I go on a wild goose chase ;)

**Comment by Lakritzator** [comment]
Actually you don't seem to be using any >.NET 4.5 features... it might be okay to leave that in for now and loose all the rest. Why build a library for many different frameworks when there is no difference?
Anyway, that would be something which @punker76 can decide on.

The only part I found which has any different "framework" code is this:
https://github.com/hardcodet/wpf-notifyicon/blob/master/Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/Interop/WindowMessageSink.cs#L188
I actually would remove this, okay? Silverlight :laughing:

**Comment by hardcodet** [comment]
@Lakritzator I'm sure Silverlight will have a massive comeback one day. Not :)

Feel free to kill that code 👍 👍 👍

File: Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/TaskbarIcon.cs (hunk 4 above)
Comment (Lakritzator) [reply Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/TaskbarIcon.cs]: Red flag... good one, I know this feeling, I have some office interop code for Greenshot where I somehow can't remove an assignment which from the compiler's point of view doesn't make any sense.

If it turns out this is something like that, a comment is needed and the discussion had value! :+1:

**Comment by hardcodet** [comment]
@Lakritzator - @punker76 I created an Azure pipeline, feel free to take over (Azure created a PR on the fly here, I'll add you guys as reviewers).

Jan, I saw that you already added a NuGet issue here, so I guess you'll know what to do there? @Lakritzator, if you'd like, I'll gladly add you to Azure as well, I'd need an email address though to send you an invite.

File: Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/TaskbarIcon.cs (hunk 4 above)
Comment (hardcodet) [reply Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/TaskbarIcon.cs]: Yup! In any case, this PR here is super helpful to me, too since my Win-Fu is quite rusty. Very happy with the conversations here, so thanks a lot for that :)

**Comment by punker76** [comment]
> Actually you don't seem to be using any >.NET 4.5 features... it might be okay to leave that in for now and loose all the rest. Why build a library for many different frameworks when there is no difference?
> Anyway, that would be something which @punker76 can decide on.
> 
> The only part I found which has any different "framework" code is this:
> https://github.com/hardcodet/wpf-notifyicon/blob/master/Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/Interop/WindowMessageSink.cs#L188
> I actually would remove this, okay? Silverlight 😆

Maybe `net40;net45;net462` and later `netcoreapp3.0` could be enough (`net462` has some fixed DPI chamges inside).

**Comment by hardcodet** [comment]
> 
> 
> > Actually you don't seem to be using any >.NET 4.5 features... it might be okay to leave that in for now and loose all the rest. Why build a library for many different frameworks when there is no difference?
> > Anyway, that would be something which @punker76 can decide on.
> > The only part I found which has any different "framework" code is this:
> > https://github.com/hardcodet/wpf-notifyicon/blob/master/Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/Interop/WindowMessageSink.cs#L188
> > I actually would remove this, okay? Silverlight 😆
> 
> Maybe `net40;net45;net462` and later `netcoreapp3.0` could be enough (`net462` has some fixed DPI chamges inside).

Sounds good! I don't see a compelling reason to continue supporting CLR 2. Unless Silverlight vnext will support it, of course :p

File: Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/TaskbarIcon.cs (hunk 4 above)
Comment (punker76) [reply Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/TaskbarIcon.cs]: I think `public GetCustomPopupPosition CustomPopupPosition { get; set; }` is good here, because the Type here is a delegate(action) which is normally set only once. So a binding makes IMO no sense.

Comment (hardcodet) [reply Hardcodet.NotifyIcon.Wpf/Source/NotifyIconWpf/TaskbarIcon.cs]: Alright then! If there's nothing you'd like to add, I'll merge this one now.
(we can adjust the build targets when we introduce the .NET Core support)

**Comment by Lakritzator** [comment]
Thanks, it has been fun! :smiley:
